### PR TITLE
Record call stack during compilation for error reporting.

### PR DESCRIPTION
**Pull request by MoFtZ**
Related to #425.

`Math.Round` does not have a Cuda intrinsic, and therefore does not currrently have a special handler in ILGPU.Algorithms. Instead, the code is dissassembled like any other normal function.

In net5.0, `Math.Round` internally calls a method in the `System.Runtime` namespace, which is not supported by ILGPU. This throws a vague exception about the method `get_IsSupported` not being supported in `Int64BitsToDouble`. However, it is not obvious where that is being called from.

This PR tracks the compilation stack to provide more detailed information about what is causing the compilation to fail. An example exception would look like this:
```
ILGPU.InternalCompilerException
  Message=
An internal compiler error has been detected
- in method Boolean get_IsSupported() declared in type System.Runtime.Intrinsics.X86.Sse2+X64
- in method Double Int64BitsToDouble(Int64) declared in type System.BitConverter
- in method Double Round(Double) declared in type System.Math
- in file 'Program.cs', line 11 column 13 to 57
- in method Void MyKernel(ILGPU.Index1, ILGPU.ArrayView`1[System.Double]) declared in type Program

Inner Exception 1:
NotSupportedException: Not supported runtime method 'get_IsSupported'
```

### CODE REVIEW

File: Src/ILGPU/IR/Location.cs [hunk 1]
```diff
@@ -246,4 +265,101 @@ public class FileLocation : Location
...
     }
+
+    /// <summary>
+    /// A collection of locations than is manipulated as a stack.
+    /// </summary>
+    public class CompilationStackLocation : Location
+    {
+        #region Instance
+
+        private ImmutableStack<Location> Stack { get; }
```
Comment (m4rs-mt) [inline Src/ILGPU/IR/Location.cs:276]: Nit: move properties (even private ones) to a `Properties` region.

**Review by m4rs-mt — APPROVED** [review]
Really nice patch 👍 This helps us to output much better error messages when compiling.

File: Src/ILGPU/IR/Location.cs (hunk 1 above)
Comment (MoFtZ) [reply Src/ILGPU/IR/Location.cs]: Fixed 👍

**Comment by m4rs-mt** [comment]
@MoFtZ looks like the latest commit causes a warning related to a protected member 😢

**Comment by MoFtZ** [comment]
Yes, I saw the build failure - wanted to fix it, but had some distractions 😄

**Pull request merged**